aresponses/main.py

```diff
@@ -80,32 +85,38 @@ def _host_matches(self, match_host):
     async def _find_response(self, request):
         host, path, path_qs, method = request.host, request.path, request.path_qs, request.method
         logger.info(f"Looking for match for {host} {path} {method}")
-        i = 0
-        host_matched = False
-        path_matched = False
-        for host_pattern, path_pattern, method_pattern, response, match_querystring in self._responses:
+        for i, (host_pattern, path_pattern, method_pattern, response, match_querystring, body) in enumerate(self._responses):
             if _text_matches_pattern(host_pattern, host):
-                host_matched = True
-                if (not match_querystring and _text_matches_pattern(path_pattern, path)) or (
-                    match_querystring and _text_matches_pattern(path_pattern, path_qs)
-                ):
-                    path_matched = True
+                if ((not match_querystring and _text_matches_pattern(path_pattern, path)) or
+                        (match_querystring and _text_matches_pattern(path_pattern, path_qs))):
                     if _text_matches_pattern(method_pattern, method.lower()):
-                        del self._responses[i]
-
-                        if callable(response):
-                            if asyncio.iscoroutinefunction(response):
-                                return await response(request)
-                            return response(request)
-
-                        if isinstance(response, str):
-                            return self.Response(body=response)
-
-                        return response
-            i += 1
-        self._exception = Exception(f"No Match found for {host} {path} {method}.  Host Match: {host_matched}  Path Match: {path_matched}")
-        self._loop.stop()
-        raise self._exception  # noqa
+                        if (isinstance(body, NoBody) or
+                                (isinstance(body, dict) and body == (await request.json())) or
+                                (isinstance(body, str) and body == (await request.text()))):
+                            self._responses.pop(i)
+
+                            if callable(response):
+                                if asyncio.iscoroutinefunction(response):
+                                    return await response(request)
+                                return response(request)
+
+                            if isinstance(response, str):
+                                return self.Response(body=response)
+
+                            return response
+        if request.body_exists:
+            try:
+                request_body = await request.json()
+            except JSONDecodeError:
+                request_body = await request.read()
+        else:
+            request_body = None
+        pytest.fail(
```

Comment:
I don't see this message show up anywhere do you?

Reply:
there is an `xfail` test, if you remove the `xfail` decorator you shall see the failing message

Reply:
i did remove it. I just get
```
____________________________________________________ test_fixture_body_json_failed _____________________________________________________
Traceback (most recent call last):
  File "/circleup/packages/aresponses/tests/test_server.py", line 76, in test_fixture_body_json_failed
    async with session.post(url, json={"a": 1}) as _:
  File "/home/bryce/.pyenv/versions/3.6.6/envs/aresponses-3.6.6/lib/python3.6/site-packages/aiohttp/client.py", line 855, in __aenter__
    self._resp = await self._coro
  File "/home/bryce/.pyenv/versions/3.6.6/envs/aresponses-3.6.6/lib/python3.6/site-packages/aiohttp/client.py", line 391, in _request
    await resp.start(conn)
  File "/home/bryce/.pyenv/versions/3.6.6/envs/aresponses-3.6.6/lib/python3.6/site-packages/aiohttp/client_reqrep.py", line 757, in start
    message, payload = await self._protocol.read()
  File "/home/bryce/.pyenv/versions/3.6.6/envs/aresponses-3.6.6/lib/python3.6/site-packages/aiohttp/streams.py", line 543, in read
    await self._waiter
aiohttp.client_exceptions.ServerDisconnectedError: None
================================================= 1 failed, 13 passed in 0.30 seconds ==================================================
```

Reply:
strange it shows up here...
```bash
(aresponses) Deos-MacBook-Pro:aresponses liangdeo$ make test
============================================================================================================ test session starts =============================================================================================================
platform darwin -- Python 3.7.3, pytest-3.6.4, py-1.8.0, pluggy-0.7.1
rootdir: /Users/liangdeo/pyproj/aresponses, inifile: tox.ini
plugins: aresponses-1.1.1, xdist-1.23.2, forked-1.0.2, cov-2.6.0, asyncio-0.9.0, pylava-0.2.1, devpi-server-4.9.0
gw0 [14] / gw1 [14] / gw2 [14] / gw3 [14] / gw4 [14] / gw5 [14] / gw6 [14] / gw7 [14]
scheduling tests via LoadScheduling
.......F......Task exception was never retrieved
future: <Task finished coro=<test_fixture_body_json_failed() done, defined at /Users/liangdeo/pyproj/aresponses/tests/test_server.py:67> exception=ServerDisconnectedError(None)>
Traceback (most recent call last):
  File "/Users/liangdeo/pyproj/aresponses/tests/test_server.py", line 74, in test_fixture_body_json_failed
    async with session.post(url, json={"a": 1}) as _:
  File "/Users/liangdeo/.virtualenvs/aresponses/lib/python3.7/site-packages/aiohttp/client.py", line 1005, in __aenter__
    self._resp = await self._coro
  File "/Users/liangdeo/.virtualenvs/aresponses/lib/python3.7/site-packages/aiohttp/client.py", line 497, in _request
    await resp.start(conn)
  File "/Users/liangdeo/.virtualenvs/aresponses/lib/python3.7/site-packages/aiohttp/client_reqrep.py", line 844, in start
    message, payload = await self._protocol.read()  # type: ignore  # noqa
  File "/Users/liangdeo/.virtualenvs/aresponses/lib/python3.7/site-packages/aiohttp/streams.py", line 588, in read
    await self._waiter
aiohttp.client_exceptions.ServerDisconnectedError: None

================================================================================================================== FAILURES ==================================================================================================================
_______________________________________________________________________________________________________ test_fixture_body_json_failed ________________________________________________________________________________________________________
[gw3] darwin -- Python 3.7.3 /Users/liangdeo/.virtualenvs/aresponses/bin/python3.7
Traceback (most recent call last):
  File "/Users/liangdeo/.virtualenvs/aresponses/lib/python3.7/site-packages/pytest_asyncio/plugin.py", line 151, in pytest_pyfunc_call
    pyfuncitem.obj(**testargs), loop=event_loop))
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 571, in run_until_complete
    self.run_forever()
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 539, in run_forever
    self._run_once()
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 1775, in _run_once
    handle._run()
  File "/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/liangdeo/pyproj/aresponses/aresponses/main.py", line 65, in _handler
    return await self._find_response(request)
  File "/Users/liangdeo/pyproj/aresponses/aresponses/main.py", line 117, in _find_response
    pytest.fail(f"No Match found for \n" f"host: {host} \n" f"path: {path} \n" f"method: {method} \n" f"body: {request_body}\n")
  File "/Users/liangdeo/.virtualenvs/aresponses/lib/python3.7/site-packages/_pytest/outcomes.py", line 97, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: No Match found for
host: foo.com
path: /
method: POST
body: {'a': 1}

==================================================================================================== 1 failed, 13 passed in 2.98 seconds =====================================================================================================
make: *** [test] Error 1
```

Reply:
interesting. we are on slightly different versions of libraries. I'll dig into it.